README
Bounced is an undocumented irc bouncer that I wrote 10 years ago.

ChangeLog

12/02/06 Colin Graf <addition@users.sourceforge.net>

	*tools.c, bounced.h:
	added vstrformat, strformat for '\0' save return
	*tools.c:
	fixed strftimet and strftimespan for '\0' save return
	*bounced.h:
	removed windows definition for snprintf, and snprintf
	increased size of OUTBUFFER which has to be greather then INBUFFER
	changed version to 0.27-dev4
	**.c:
	changed use of snprintf to use of strformat
	fixed use of strftimet and strftimetspan
	*sample.motd:
	updated this file
	*motd.c:
	fixed value of some motd vars

01/21/06 Colin Graf <addition@users.sourceforge.net>

	*mode.c:
	fixed some mode attaching bugs

01/16/06 Colin Graf <addition@users.sourceforge.net>

	*dmemory.c, dmemory.h:
	added function DebugMemoryCheckBlocks() for detecting "brocken" memory... (brocken = where there blockheader changed... )
	reduced DEBUG_MEMORY_BLOCK_HASH_SIZE to 1024+1
	*connection.c, bounced.c:
	added DebugMemoryCheckBlocks() call after cmd handler and at main loop's end

01/09/06 Colin Graf <addition@users.sourceforge.net>

	*nick.c:
	fixed new NICK request, when the current client nick is not equal the server nick

01/02/06 Colin Graf <addition@users.sourceforge.net>

	*nick.c:
	fixed prefered nick reusing bug

05/04/05 Colin Graf <addition@users.sourceforge.net>

	*profile.c:
	fixed bug where the bouncer doesn't free the away reason on profile attach
	*ping.c:
	change profile detach on ping timeout code

03/31/05 Colin Graf <addition@users.sourceforge.net>

	*profile.c:
	change "profile" irccmd usage description
	*config.c:
	change "admin config" irccmd usage description
	*admin.c:
	fixed "privmsg", "notice", "wallops" irccmd usage description

02/17/05 Colin Graf <addition@users.sourceforge.net>

	*connection.c, bounced.c:
	changed invalid socket detection after socket() or accept() call

02/15/05 Colin Graf <addition@users.sourceforge.net>

	*connection.c:
	fixed bug in ConnectionRead()
	fixed bug in ConnectionWrite()
	improved ConnectionConnect()

02/14/05 Colin Graf <addition@users.sourceforge.net>

	*tools.c:
	improved Log() function
	*connection.c:
	fixed bug in ConnectionSend()

version 0.26 - 01/12/05 Colin Graf <addition@users.sourceforge.net>

	*dbaccess.c:
	added profile connect/disconnect commads to dbaccess stuff
	*bounced.c:
	added a ConnectionsClose() call before entering select loop

01/11/05 Colin Graf <addition@users.sourceforge.net>

	*user.c, dbaccess.c:
	added some "g_bUsersChanged = 1;"
	*connect.c, server.c, user.c, admin.c, dbaccess.c, bounced.h:
	added special "ConnectInterface" setting for each user
	*user.c:
	added "/admin user email" inIRC command

09/08/04 Colin Graf <addition@users.sourceforge.net>

	*dbaccess.c:
	changed some ConnectionSendFormat to ConnectionSend
	*profile.c:
	now an "online nick" will be set to pServer->strNick when a profile gets attached while it is connecting (hope this works)

version 0.25 - 08/26/04 Colin Graf <addition@users.sourceforge.net>

08/23/04 Colin Graf <addition@users.sourceforge.net>

	*tools.c:
	added check if strNick from CompareIRCAddressNick is empty
	*server.c:
	added check if SolveHostname was successful to avoid senceless connect tries
	*profile.c:
	added g_bShutdown check to ProfileConnectTimer
	fixed error handling bug in ProfileConnect if ServerConnect() failes
	*connect.c
	fixed small error handling bug in ConnectionConnect if socket() failes

08/18/04 Colin Graf <addition@users.sourceforge.net>

	*dbaccess.c:
	changed value of "server" in "profile get"

08/16/04 Colin Graf <addition@users.sourceforge.net>

	*dbaccess.c:
	fixed "profiles" and "clients" from "admin user get"
	fixed bug in "statistic get"

version 0.24 - 08/11/04 Colin Graf <addition@users.sourceforge.net>

	*user.c:
	changed UserSetAllowedCommands
 
08/10/04 Colin Graf <addition@users.sourceforge.net>

	*dbaccess.c:
	fixed bug in "config get"
	added "server" command
	*profile.c:
	added CV_READONLY to "Servers"
	*server.c:
	added "server set"

version 0.23 - 07/30/04 Colin Graf <addition@users.sourceforge.net>

	*user.c:
	added "/admin user commands" command

07/27/04 Colin Graf <addition@users.sourceforge.net>

	*server.c, profile.c:
	improved autoconnect stuff
	*config.c:
	added support for "/config set <var>" without <value>
	added support for "/admin config set <var>" without <value>
	*dbaccess.c:
	added "admin user commands" command

07/25/04 Colin Graf <addition@users.sourceforge.net>

	*nick.c:
	changed success return value of ServerHandlerUnregisteredNickError to 0 (now clients can see the error message)
	*bounced.h, bounced.c, config.c, user.c, profile.c, ...
	changed flags stuff and removed RestrictCommands (now users have specific AllowedCommands)

07/23/04 Colin Graf <addition@users.sourceforge.net>

	*quit.c:
	fixed nick reusing bug

version 0.22 - 07/12/04 Colin Graf <addition@users.sourceforge.net>

06/21/04 Colin Graf <addition@users.sourceforge.net>

	*tools.c, bounced.c, bounced.h:
	improved unix log file stuff
	*profile.c
	fixed help message bug in ClientHandlerProfile

06/20/04 Colin Graf <addition@users.sourceforge.net>

	*profile.c:
	added profile name check to /profile add

	*motd.c, bounced.c, bounced.h:
	added g_strSystem
	changed g_strMotdSystem stuff

06/18/04 Colin Graf <addition@users.sourceforge.net>

	*dbaccess.c:
	extended dbaccess command

06/17/04 Colin Graf <addition@users.sourceforge.net>

	*config.c:
	added CV_HIDDEN to DBAccessPassword

version 0.21 - 06/17/04 Colin Graf <addition@users.sourceforge.net>

06/09/04 Colin Graf <addition@users.sourceforge.net> 

	*user.c:
	improved profile status in profile list (added "connecting")

05/01/04 Colin Graf <addition@users.sourceforge.net> 

	*user.c:
	fixed profile status in profile list

04/29/04 Colin Graf <addition@users.sourceforge.net> 

	*privmsg.c:
	added not logging of ctcp answers to ServerHandlerNotice
	*bounced.h:
	changed DEFAULT_NICKLEN to 15
	*nick.c, bounced.h:
	improved NickIsValid (rfc specifications)


src/md5.c
/* md5.c - Functions to compute MD5 message digest of files or memory blocks
   according to the definition of MD5 in RFC 1321 from April 1992.
   Copyright (C) 1995, 1996 Free Software Foundation, Inc.
   NOTE: The canonical source of this file is maintained with the GNU C
   Library.  Bugs can be reported to bug-glibc@prep.ai.mit.edu.

   This program is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by the
   Free Software Foundation; either version 2, or (at your option) any
   later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */

/* Written by Ulrich Drepper <drepper@gnu.ai.mit.edu>, 1995.  */

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#if defined(_WIN32) && !defined(__CYGWIN__) && !defined(STDC_HEADERS)
#define STDC_HEADERS 1
#endif

#include <sys/types.h>

#if STDC_HEADERS || defined _LIBC
# include <stdlib.h>
# include <string.h>
#else
# ifndef HAVE_MEMCPY
#  define memcpy(d, s, n) bcopy ((s), (d), (n))
# endif
#endif

#include "md5.h"

#ifdef _LIBC
# include <endian.h>
# if __BYTE_ORDER == __BIG_ENDIAN
#  define WORDS_BIGENDIAN 1
# endif
#endif

#ifdef WORDS_BIGENDIAN
# define SWAP(n)              \
    (((n) << 24) | (((n) & 0xff00) << 8) | (((n) >> 8) & 0xff00) | ((n) >> 24))
#else
# define SWAP(n) (n)
#endif


/* This array contains the bytes used to pad the buffer to the next
   64-byte boundary.  (RFC 1321, 3.1: Step 1)  */
static const unsigned char fillbuf[64] = { 0x80, 0 /* , 0, 0, ...  */  };


/* Initialize structure containing state of computation.
   (RFC 1321, 3.3: Step 3)  */
void
md5_init_ctx (ctx)
     struct md5_ctx *ctx;
{
    ctx->A = 0x67452301;
    ctx->B = 0xefcdab89;
    ctx->C = 0x98badcfe;
    ctx->D = 0x10325476;

    ctx->total[0] = ctx->total[1] = 0;
    ctx->buflen = 0;
}

/* Put result from CTX in first 16 bytes following RESBUF.  The result
   must be in little endian byte order.

   IMPORTANT: On some systems it is required that RESBUF is correctly
   aligned for a 32 bits value.  */
void *
md5_read_ctx (ctx, resbuf)
     const struct md5_ctx *ctx;
     void *resbuf;
{
    ((md5_uint32 *) resbuf)[0] = SWAP (ctx->A);
    ((md5_uint32 *) resbuf)[1] = SWAP (ctx->B);
    ((md5_uint32 *) resbuf)[2] = SWAP (ctx->C);
    ((md5_uint32 *) resbuf)[3] = SWAP (ctx->D);

    return resbuf;
}

/* Process the remaining bytes in the internal buffer and the usual
   prolog according to the standard and write the result to RESBUF.

   IMPORTANT: On some systems it is required that RESBUF is correctly
   aligned for a 32 bits value.  */
void *
md5_finish_ctx (ctx, resbuf)
     struct md5_ctx *ctx;
     void *resbuf;
{
    /* Take yet unprocessed bytes into account.  */
    md5_uint32 bytes = ctx->buflen;
    size_t pad;

    /* Now count remaining bytes.  */
    ctx->total[0] += bytes;
    if (ctx->total[0] < bytes)
  ++ctx->total[1];

    pad = bytes >= 56 ? 64 + 56 - bytes : 56 - bytes;
    memcpy (&ctx->buffer[bytes], fillbuf, pad);

    /* Put the 64-bit file length in *bits* at the end of the buffer.  */
    *(md5_uint32 *) & ctx->buffer[bytes + pad] = SWAP (ctx->total[0] << 3);
    *(md5_uint32 *) & ctx->buffer[bytes + pad + 4] =
  SWAP ((ctx->total[1] << 3) | (ctx->total[0] >> 29));

    /* Process last bytes.  */
    md5_process_block (ctx->buffer, bytes + pad + 8, ctx);

    return md5_read_ctx (ctx, resbuf);
}

/* Compute MD5 message digest for bytes read from STREAM.  The
   resulting message digest number will be written into the 16 bytes
   beginning at RESBLOCK.  */
int
md5_stream (stream, resblock)
     FILE *stream;
     void *resblock;
{
    /* Important: BLOCKSIZE must be a multiple of 64.  */
#define BLOCKSIZE 4096
    struct md5_ctx ctx;
    char buffer[BLOCKSIZE + 72];
    size_t sum;

    /* Initialize the computation context.  */
    md5_init_ctx (&ctx);

    /* Iterate over full file contents.  */
    while (1)
    {
  /* We read the file in blocks of BLOCKSIZE bytes.  One call of the
     computation function processes the whole buffer so that with the
     next round of the loop another block can be read.  */
  size_t n;

  sum = 0;

  /* Read block.  Take care for partial reads.  */
  do
  {
      n = fread (buffer + sum, 1, BLOCKSIZE - sum, stream);

      sum += n;
  }
  while (sum < BLOCKSIZE && n != 0);
  if (n == 0 && ferror (stream))
      return 1;

  /* If end of file is reached, end the loop.  */
  if (n == 0)
      break;

  /* Process buffer with BLOCKSIZE bytes.  Note that
     BLOCKSIZE % 64 == 0
   */
  md5_process_block (buffer, BLOCKSIZE, &ctx);
    }

    /* Add the last bytes if necessary.  */
    if (sum > 0)
  md5_process_bytes (buffer, sum, &ctx);

    /* Construct result in desired memory.  */
    md5_finish_ctx (&ctx, resblock);
    return 0;
}

/* Compute MD5 message digest for LEN bytes beginning at BUFFER.  The
   result is always in little endian byte order, so that a byte-wise
   output yields to the wanted ASCII representation of the message
   digest.  */
void *
md5_buffer (buffer, len, resblock)
     const char *buffer;
     size_t len;
     void *resblock;
{
    struct md5_ctx ctx;

    /* Initialize the computation context.  */
    md5_init_ctx (&ctx);

    /* Process whole buffer but last len % 64 bytes.  */
    md5_process_bytes (buffer, len, &ctx);

    /* Put result in desired memory area.  */
    return md5_finish_ctx (&ctx, resblock);
}


void
md5_process_bytes (buffer, len, ctx)
     const void *buffer;
     size_t len;
     struct md5_ctx *ctx;
{
    /* When we already have some bits in our internal buffer concatenate
       both inputs first.  */
    if (ctx->buflen != 0)
    {
  size_t left_over = ctx->buflen;
  size_t add = 128 - left_over > len ? len : 128 - left_over;

  memcpy (&ctx->buffer[left_over], buffer, add);
  ctx->buflen += add;

  if (left_over + add > 64)
  {
      md5_process_block (ctx->buffer, (left_over + add) & ~63, ctx);
      /* The regions in the following copy operation cannot overlap.  */
      memcpy (ctx->buffer, &ctx->buffer[(left_over + add) & ~63],
        (left_over + add) & 63);
      ctx->buflen = (left_over + add) & 63;
  }

  buffer = (const char *) buffer + add;
  len -= add;
    }

    /* Process available complete blocks.  */
    if (len > 64)
    {
  md5_process_block (buffer, len & ~63, ctx);
  buffer = (const char *) buffer + (len & ~63);
  len &= 63;
    }

    /* Move remaining bytes in internal buffer.  */
    if (len > 0)
    {
  memcpy (ctx->buffer, buffer, len);
  ctx->buflen = len;
    }
}


/* These are the four functions used in the four steps of the MD5 algorithm
   and defined in the RFC 1321.  The first function is a little bit optimized
   (as found in Colin Plumbs public domain implementation).  */
/* #define FF(b, c, d) ((b & c) | (~b & d)) */
#define FF(b, c, d) (d ^ (b & (c ^ d)))
#define FG(b, c, d) FF (d, b, c)
#define FH(b, c, d) (b ^ c ^ d)
#define FI(b, c, d) (c ^ (b | ~d))

/* Process LEN bytes of BUFFER, accumulating context into CTX.
   It is assumed that LEN % 64 == 0.  */

void
md5_process_block (buffer, len, ctx)
     const void *buffer;
     size_t len;
     struct md5_ctx *ctx;
{
    md5_uint32 correct_words[16];
    const md5_uint32 *words = buffer;
    size_t nwords = len / sizeof (md5_uint32);
    const md5_uint32 *endp = words + nwords;
    md5_uint32 A = ctx->A;
    md5_uint32 B = ctx->B;
    md5_uint32 C = ctx->C;
    md5_uint32 D = ctx->D;

    /* First increment the byte count.  RFC 1321 specifies the possible
       length of the file up to 2^64 bits.  Here we only compute the
       number of bytes.  Do a double word increment.  */
    ctx->total[0] += len;
    if (ctx->total[0] < len)
  ++ctx->total[1];

    /* Process all bytes in the buffer with 64 bytes in each round of
       the loop.  */
    while (words < endp)
    {
  md5_uint32 *cwp = correct_words;
  md5_uint32 A_save = A;
  md5_uint32 B_save = B;
  md5_uint32 C_save = C;
  md5_uint32 D_save = D;

  /* First round: using the given function, the context and a constant
     the next context is computed.  Because the algorithms processing
     unit is a 32-bit word and it is determined to work on words in
     little endian byte order we perhaps have to change the byte order
     before the computation.  To reduce the work for the next steps
     we store the swapped words in the array CORRECT_WORDS.  */

#define OP(a, b, c, d, s, T)            \
      do                \
        {                \
    a += FF (b, c, d) + (*cwp++ = SWAP (*words)) + T;    \
    ++words;              \
    CYCLIC (a, s);            \
    a += b;              \
        }                \
      while (0)

  /* It is unfortunate that C does not provide an operator for
     cyclic rotation.  Hope the C compiler is smart enough.  */
#define CYCLIC(w, s) (w = (w << s) | (w >> (32 - s)))

  /* Before we start, one word to the strange constants.
     They are defined in RFC 1321 as

     T[i] = (int) (4294967296.0 * fabs (sin (i))), i=1..64
   */

  /* Round 1.  */
  OP (A, B, C, D, 7, 0xd76aa478);
  OP (D, A, B, C, 12, 0xe8c7b756);
  OP (C, D, A, B, 17, 0x242070db);
  OP (B, C, D, A, 22, 0xc1bdceee);
  OP (A, B, C, D, 7, 0xf57c0faf);
  OP (D, A, B, C, 12, 0x4787c62a);
  OP (C, D, A, B, 17, 0xa8304613);
  OP (B, C, D, A, 22, 0xfd469501);
  OP (A, B, C, D, 7, 0x698098d8);
  OP (D, A, B, C, 12, 0x8b44f7af);
  OP (C, D, A, B, 17, 0xffff5bb1);
  OP (B, C, D, A, 22, 0x895cd7be);
  OP (A, B, C, D, 7, 0x6b901122);
  OP (D, A, B, C, 12, 0xfd987193);
  OP (C, D, A, B, 17, 0xa679438e);
  OP (B, C, D, A, 22, 0x49b40821);

  /* For the second to fourth round we have the possibly swapped words
     in CORRECT_WORDS.  Redefine the macro to take an additional first
     argument specifying the function to use.  */
#undef OP
#define OP(f, a, b, c, d, k, s, T)          \
      do                 \
  {                \
    a += f (b, c, d) + correct_words[k] + T;      \
    CYCLIC (a, s);            \
    a += b;              \
  }                \
      while (0)

  /* Round 2.  */
  OP (FG, A, B, C, D, 1, 5, 0xf61e2562);
  OP (FG, D, A, B, C, 6, 9, 0xc040b340);
  OP (FG, C, D, A, B, 11, 14, 0x265e5a51);
  OP (FG, B, C, D, A, 0, 20, 0xe9b6c7aa);
  OP (FG, A, B, C, D, 5, 5, 0xd62f105d);
  OP (FG, D, A, B, C, 10, 9, 0x02441453);
  OP (FG, C, D, A, B, 15, 14, 0xd8a1e681);
  OP (FG, B, C, D, A, 4, 20, 0xe7d3fbc8);
  OP (FG, A, B, C, D, 9, 5, 0x21e1cde6);
  OP (FG, D, A, B, C, 14, 9, 0xc33707d6);
  OP (FG, C, D, A, B, 3, 14, 0xf4d50d87);
  OP (FG, B, C, D, A, 8, 20, 0x455a14ed);
  OP (FG, A, B, C, D, 13, 5, 0xa9e3e905);
  OP (FG, D, A, B, C, 2, 9, 0xfcefa3f8);
  OP (FG, C, D, A, B, 7, 14, 0x676f02d9);
  OP (FG, B, C, D, A, 12, 20, 0x8d2a4c8a);

  /* Round 3.  */
  OP (FH, A, B, C, D, 5, 4, 0xfffa3942);
  OP (FH, D, A, B, C, 8, 11, 0x8771f681);
  OP (FH, C, D, A, B, 11, 16, 0x6d9d6122);
  OP (FH, B, C, D, A, 14, 23, 0xfde5380c);
  OP (FH, A, B, C, D, 1, 4, 0xa4beea44);
  OP (FH, D, A, B, C, 4, 11, 0x4bdecfa9);
  OP (FH, C, D, A, B, 7, 16, 0xf6bb4b60);
  OP (FH, B, C, D, A, 10, 23, 0xbebfbc70);
  OP (FH, A, B, C, D, 13, 4, 0x289b7ec6);
  OP (FH, D, A, B, C, 0, 11, 0xeaa127fa);
  OP (FH, C, D, A, B, 3, 16, 0xd4ef3085);
  OP (FH, B, C, D, A, 6, 23, 0x04881d05);
  OP (FH, A, B, C, D, 9, 4, 0xd9d4d039);
  OP (FH, D, A, B, C, 12, 11, 0xe6db99e5);
  OP (FH, C, D, A, B, 15, 16, 0x1fa27cf8);
  OP (FH, B, C, D, A, 2, 23, 0xc4ac5665);

  /* Round 4.  */
  OP (FI, A, B, C, D, 0, 6, 0xf4292244);
  OP (FI, D, A, B, C, 7, 10, 0x432aff97);
  OP (FI, C, D, A, B, 14, 15, 0xab9423a7);
  OP (FI, B, C, D, A, 5, 21, 0xfc93a039);
  OP (FI, A, B, C, D, 12, 6, 0x655b59c3);
  OP (FI, D, A, B, C, 3, 10, 0x8f0ccc92);
  OP (FI, C, D, A, B, 10, 15, 0xffeff47d);
  OP (FI, B, C, D, A, 1, 21, 0x85845dd1);
  OP (FI, A, B, C, D, 8, 6, 0x6fa87e4f);
  OP (FI, D, A, B, C, 15, 10, 0xfe2ce6e0);
  OP (FI, C, D, A, B, 6, 15, 0xa3014314);
  OP (FI, B, C, D, A, 13, 21, 0x4e0811a1);
  OP (FI, A, B, C, D, 4, 6, 0xf7537e82);
  OP (FI, D, A, B, C, 11, 10, 0xbd3af235);
  OP (FI, C, D, A, B, 2, 15, 0x2ad7d2bb);
  OP (FI, B, C, D, A, 9, 21, 0xeb86d391);

  /* Add the starting values of the context.  */
  A += A_save;
  B += B_save;
  C += C_save;
  D += D_save;
    }

    /* Put checksum in context given as argument.  */
    ctx->A = A;
    ctx->B = B;
    ctx->C = C;
    ctx->D = D;
}
